src/System.Linq.Expressions/src/System/Runtime/CompilerServices/CallSite.cs

```diff
@@ -667,7 +667,7 @@ private T CreateCustomNoMatchDelegate(MethodInfo invoke)
                 Expression.Block(
                     Expression.Call(
                         typeof(CallSiteOps).GetMethod(nameof(CallSiteOps.SetNotMatched)),
-                        @params.First()
+                        @params[0]
```

Comment:
I think assertion would not be very useful. 
Many callsite members and generated helpers expect signatures of a particular shape, but they are mostly pieces that call each other and are not supposed to be called externally.